### ISSUE TITLE
SpringViewProvider throws ConcurrentModificationException

### ISSUE THREAD

**Issue by andreabreu-me** [vaadin/spring#9]
I am getting ConcurrentModificationException when I try to get views from the SpringViewProvider. The strange part is that it only throws the exception when I try to get the last view.

```java
@SpringBootApplication
public class DemoApplication {

    public static void main(String[] args) {
        SpringApplication.run(DemoApplication.class, args);
    }

    @SpringUI(path = "/ui")
    @Theme("valo")
    public static class MainUI extends UI {

        @Autowired
        SpringViewProvider viewProvider;

        @Override
        protected void init(VaadinRequest request) {
            final Navigator navigator = new Navigator(this, this);
            navigator.addProvider(viewProvider);
            navigator.setErrorView(new ErrorView());
            setNavigator(navigator);

            View a = viewProvider.getView("foo");
            View b = viewProvider.getView("bar");
            View c = viewProvider.getView("hello");
            View d = viewProvider.getView("world");
        }
    }

    private static abstract class AbstactView extends CssLayout implements View {

        public AbstactView() {
            addComponent(new Label(this.getClass().getName()));
        }

        @Override
        public void enter(ViewChangeListener.ViewChangeEvent event) {
        }
    }

    @SpringView(name = "foo")
    @ViewScope
    private static class FooView extends AbstactView {}

    @SpringView(name = "bar")
    @ViewScope
    private static class BarView extends AbstactView {}

    @SpringView(name = "hello")
    @ViewScope
    private static class HelloView extends AbstactView {}

    @SpringView(name = "world")
    @ViewScope
    private static class WorldView extends AbstactView {}

    @SpringView(name = "error")
    @ViewScope
    private static class ErrorView extends AbstactView {}
}
```

Log:
```java
java.util.ConcurrentModificationException: null
	at java.util.LinkedList$ListItr.checkForComodification(LinkedList.java:966)
	at java.util.LinkedList$ListItr.next(LinkedList.java:888)
	at com.vaadin.navigator.Navigator.fireAfterViewChange(Navigator.java:650)
	at com.vaadin.navigator.Navigator.navigateTo(Navigator.java:580)
	at com.vaadin.navigator.Navigator.navigateTo(Navigator.java:526)
	at com.vaadin.ui.UI.doInit(UI.java:651)
	at com.vaadin.server.communication.UIInitHandler.getBrowserDetailsUI(UIInitHandler.java:214)
	at com.vaadin.server.communication.UIInitHandler.synchronizedHandleRequest(UIInitHandler.java:74)
	at com.vaadin.server.SynchronizedRequestHandler.handleRequest(SynchronizedRequestHandler.java:41)
	at com.vaadin.server.VaadinService.handleRequest(VaadinService.java:1408)
	at com.vaadin.server.VaadinServlet.service(VaadinServlet.java:350)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:725)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:291)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.springframework.web.filter.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:77)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:85)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:219)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:106)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:501)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:142)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:79)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:88)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:516)
	at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1086)
	at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:659)
	at org.apache.coyote.http11.Http11NioProtocol$Http11ConnectionHandler.process(Http11NioProtocol.java:223)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1558)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1515)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by hesara** [comment]
This looks like the issue originally reported as https://dev.vaadin.com/ticket/17475, which turned out to be the Vaadin Navigator issue https://dev.vaadin.com/ticket/17477. It should be fixed in Vaadin 7.4.4 (out very soon), so please re-test with it once it is released and close this issue if fixed by Vaadin 7.4.4.

**Comment by andreabreu-me** [comment]
The issue still exists in Vaadin 7.4.4

**Comment by hesara** [comment]
Sorry, it looks like the fix only ended up in the master branch and not in 7.4.4 - I'll try to get it included in 7.4.5.

**Comment by samie** [comment]
This was fixed in Vaadin 7.4.5. Simple test case here: https://github.com/samie/test-springviewprovider
